src/placement/pl_map_common.c

```diff
@@ -233,51 +237,58 @@ remap_list_fill(struct pl_map *map, struct daos_obj_md *md,
 }
 
 bool
-need_remap_target(struct pool_target *tgt, uint32_t allow_status, uint32_t allow_version)
+need_remap_comp(struct pool_component *comp, uint32_t allow_status, uint32_t allow_version,
+		bool for_reint)
 {
-	if (!pool_target_avail(tgt, allow_status))
-		return true;
-
-	if (tgt->ta_comp.co_status == PO_COMP_ST_UPIN || allow_version == (uint32_t)(-1))
-		return false;
-
-	if (tgt->ta_comp.co_status == PO_COMP_ST_DRAIN) {
-		/* If Drain happenes later, let's ignore it for current rebuild/reintegration */
-		if (tgt->ta_comp.co_fseq > allow_version)
-			return false;
-		else
-			return true;
+	uint32_t status = comp->co_status;
+
+	/* Correct the target status by allow_version */
+	if (status == PO_COMP_ST_UP) {
+		if (comp->co_in_ver > allow_version) {
+			if (comp->co_fseq > 0)
+				status = PO_COMP_ST_DOWNOUT;
+			else
+				status = PO_COMP_ST_NEW;
+		}
+	} else if (for_reint && (status == PO_COMP_ST_DOWN)) {
+		/* In jump_map_obj_find_reint(), it needs to compare between the original layout
+		 * and new(after extend/reintegration) layout to figure out which shard needs to
+		 * be relocated, so it has to follow the failure sequence strictly, i.e. all the
+		 * the target has to be restored to its original status.
+		 * In other cases, it has to skip the DOWN target in all cases, otherwise the new
+		 * generated layout might include the failure shards, and inflight I/O might hit
+		 * the failure rank.
+		 */
+		if (comp->co_fseq > allow_version)
+			status = PO_COMP_ST_UPIN;
```

Comment:
Hmm, indeed. but there are not enough information to restore its exact status, i.e. these information is not stored in the pool map. 

Though this is only for checking if target is available during reintegration process. So it does not matter if we restore to DRAIN or UPIN, since both of them are available in that case anyway.